Add n_components="auto" to NMF when H and W are provided

Alexlandeau opened the pull request:
Fixes #26392 
* adding "auto" as default value for n_components. In this mode, n_components will be inferred from H and W if they are provided
* adding additional validation on inferred n_components to prevent inconsistencies
* updating parameter validation, deprecation notice, documentation and tests accordingly

Alexlandeau committed:
Fix issue 26392 on NMF

* adding "auto" as default value for n_components
* updating parameter validation, deprecation notice, documentation and tests accordingly
---------

Co-authored-by: Alexandre Landeau <alandeau@MacBook-Pro-de-Alexandre.local>
Co-authored-by: avigny <alexandre.vigny@dataiku.com>

Commit:
fixed linting

Commit:
Merge branch 'main' into bug/nmf_26392

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1183,6 +1201,19 @@ def _check_params(self, X):
+            else:
+                # Default to n_features
+                self._n_components = X.shape[1]
+
         """Check W and H, or initialize them."""
```
the docstring should stay at the top of the function

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1183,6 +1201,19 @@ def _check_params(self, X):
+        if self._n_components == "auto":
+            if W is not None:
+                if H is not None:
+                    self._n_components = H.shape[0]
```
Here I would raise an error if shapes don't match (H.shape[0] and W.shape[1])

Alexlandeau replied:
added in [261e5404ac63961856c628ef82c34f4979eea018](https://github.com/scikit-learn/scikit-learn/pull/26634/commits/261e5404ac63961856c628ef82c34f4979eea018)

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1183,6 +1201,19 @@ def _check_params(self, X):
+            if W is not None:
+                if H is not None:
+                    self._n_components = H.shape[0]
+                else:
+                    self._n_components = W.shape[1]
```
I don't think this case can happen. We can have either W and H provided or only H provided but not only W

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1349,7 +1378,7 @@ class NMF(_BaseNMF):
     Parameters
     ----------
-    n_components : int, default=None
+    n_components : int or {'auto'} or None, default=None
```
please add a `versionchanged` directive to say that "auto" option was added in 1.4

jeremiedbb replied:
same for MiniBatchNMF

avigny replied:
done in 1c6994be8fdf7693711e5a59630101b4aee18914

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -973,9 +973,14 @@ def non_negative_factorization(
+        If `n_components='auto'`, the number of components is automatically infered
+        from W or H shapes.
+
+        .. versionchanged:: 1.3
+           The default value of `n_components` will change to "auto" in v1.5.
```
instead there should be a `versionchanged` directive saying that the "auto" option was added in 1.4

avigny replied:
done in 1c6994be8fdf7693711e5a59630101b4aee18914

jeremiedbb commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -933,3 +933,29 @@ def test_NMF_inverse_transform_W_deprecation():
         est.inverse_transform(W=Xt)
+
+
+def test_NMF_n_components_auto():
```
Please write similar tests for NMF and MiniBatchNMF

avigny replied:
done in af565549c1339ed2fd3061c0b5b4532af2682b46
https://github.com/scikit-learn/scikit-learn/blob/af565549c1339ed2fd3061c0b5b4532af2682b46/sklearn/decomposition/tests/test_nmf.py#L938-L951

jeremiedbb commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -933,3 +933,29 @@ def test_NMF_inverse_transform_W_deprecation():
+    W, H, _ = non_negative_factorization(X, n_components="auto")
+    assert W.shape[1] == X.shape[1] and H.shape[0] == X.shape[1]
+
+    # H_true is set, inferring n_components from it
+    W, H, _ = non_negative_factorization(X, H=H_true, n_components="auto")
+    assert H.shape[0] == H_true.shape[0]
+
+    # H_true is set, inferring n_components from it
+    W, H, _ = non_negative_factorization(X, W=W_true, n_components="auto")
+    assert W.shape[1] == W_true.shape[1]
```
This is not testing what you expect. Since you didn't set `init="custom"`, W and H are ignored.

Maybe we should raise if ``init!="custom"`` and W or H is passed.

avigny replied:
I would say that W and H are ignored if you don't set `init="custom"` or you don't set `update_H=False`

What we could do (but I'll let you decide if this is in or out of the scope of this PR) is top do a check here
https://github.com/scikit-learn/scikit-learn/blob/0e1f170d8a392c6c43a67edf0d7f46150956aacf/sklearn/decomposition/_nmf.py#L1213-L1216
and warn the user if `W` or `H` was provided and will not be used (because they are initialised regardless)

jeremiedbb replied:
A warning is good, yes. I'd do that in this PR.

We also need a warning for W only just above, when update_H is False and W is not None.

avigny replied:
done in https://github.com/scikit-learn/scikit-learn/pull/26634/commits/82fbc4c064481d573199fdd0dc85e76359f382f0

jeremiedbb reviewed (commented):
Thanks for the PR @Alexlandeau.

Here are some comments. In particular, this pr should target version 1.4 since the feature freeze for 1.3 already happened. Also, the addition of `auto` and future change of default should also be done for NMF and MiniBatchNMF.

Please also add an entry in the `v1.4.rst` changelog.

Commit:
linting

Alexlandeau committed:
Apply suggestions from code review

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

avigny committed:
import _param_validation.hidden and close parenthesis

avigny committed:
Doc for fit_transform of MiniBatchNMF and NMF classes updated

avigny committed:
Adding `versionchanged` directive about ading `'auto`` value

avigny committed:
Moving the `n_components` validation in the `_check_params` function

avigny commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1097,6 +1102,15 @@ def non_negative_factorization(
+            "`None` to `'auto'` in 1.5.",
+            FutureWarning,
+        )
+        n_components = None
```
Done in https://github.com/scikit-learn/scikit-learn/pull/26634/commits/9f031da133b1048ca208a5191b07f65369eef6c2
I moved this to the `_check_params` function since it appears this is where the checks are done.
https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1250

Added bonus with this: the check only needs to be done in the base class `_BaseNMF` as it is called from the child classes:
- `NMF`: https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1548
- `MiniBatchNMF`: https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1992

Alexlandeau commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1183,6 +1201,19 @@ def _check_params(self, X):
 
     def _check_w_h(self, X, W, H, update_H):
+        # n_components
+        if self._n_components == "auto":
+            if W is not None:
+                if H is not None:
+                    self._n_components = H.shape[0]
+                else:
+                    self._n_components = W.shape[1]
+            elif H is not None:
+                self._n_components = H.shape[0]
+            else:
+                # Default to n_features
+                self._n_components = X.shape[1]
```
as discussed IRL, we can't move it at the end as we need to initialize `_n_components` before the `_check_init` call, where W and H shapes are checked. Will replace the `.shape` calls by corresponding `_num_features` and `_num_samples` calls, in order to make this initialization work on array-like objects as well

Commit:
fixed _n_components initialization when init method is custom

Commit:
added additional validation on W and H shapes when provided

avigny committed:
Update tests
adding `init="custom"` so that we actually test the modified code

Alexlandeau committed:
Merge branch 'main' into bug/nmf_26392

Commit:
merge conflicts

Commit:
merge conflicts pt.2

Alexlandeau committed:
Merge branch 'main' into bug/nmf_26392

Commit:
added changelog

Commit:
Merge branch 'bug/nmf_26392' of github.com:Alexlandeau/scikit-learn into bug/nmf_26392

avigny committed:
Allow `n_components="auto"` independently of the init method.
This allows to check the shapes of w or h if given to the estimator.
added the corresponding test as well

jeremiedbb commented on doc/whats_new/v1.4.rst:
```diff
@@ -58,3 +58,6 @@ TODO: update at the time of the release.
   :meth:`base.OutlierMixin.fit_predict` now accept ``**kwargs`` which are
   passed to the ``fit`` method of the the estimator. :pr:`26506` by `Adrin
   Jalali`_.
+- |Enhancement| The default value in the `n_components` parameter of :func:`non_negative_factorization` in :class:`decomposition.NMF` and
+  :class:`decomposition.MiniBatchNMF` will change from `None` to `auto` in version 1.6, so that its value is automatically computed when H is provided.
+  :pr:`26634` by :user:`Alexandre Landeau <AlexL>` and :user:`Alexandre Vigny <avigny>`.
```
```suggestion

:mod:`sklearn.decomposition`
............................

- |Enhancement| An "auto" option was added to the `n_components` parameter of
  :func:`decomposition.non_negative_factorization`, :class:`decomposition.NMF` and
  :class:`decomposition.MiniBatchNMF` to automatically infer the number of components
  when using a custom initialization. The default value of this parameter will change
  from `None` to `auto` in version 1.6.
  :pr:`26634` by :user:`Alexandre Landeau <AlexL>` and :user:`Alexandre Vigny <avigny>`.
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -976,9 +979,15 @@ def non_negative_factorization(
+    n_components : int or {'auto'} or None, default=None
         Number of components, if n_components is not set all features
         are kept.
+        If `n_components='auto'`, the number of components is automatically infered
+        from W or H shapes.
```
```suggestion
        If `n_components='auto'`, the number of components is automatically inferred
        from W or H shapes.
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1179,7 +1194,19 @@ def __init__(
+                (
+                    "The default value of `n_components` will change from `None` to"
+                    " `'auto'` in 1.6. Set the value of `n_components` to `None`"
+                    " explicitly to supress the warning"
```
```suggestion
                    " explicitly to supress the warning."
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1179,7 +1194,19 @@ def __init__(
+            )
+            self._n_components = None  # Keeping the old default value
+        if self._n_components is None or (
+            self.init != "custom" and self.n_components == "auto"
```
```suggestion
            self.init != "custom" and self._n_components == "auto"
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1188,6 +1215,24 @@ def _check_params(self, X):
     def _check_w_h(self, X, W, H, update_H):
         """Check W and H, or initialize them."""
         n_samples, n_features = X.shape
+        if self.n_components == "auto":
```
```suggestion
        if self._n_components == "auto":
```

Alexlandeau replied:
I think here we shouldn't replace it by _n_components here as _n_components is set to X.shape[1] a few lines before, in _check_params, when "auto" is selected. So that statement would never be true

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1188,6 +1215,24 @@ def _check_params(self, X):
+            else:
+                self._n_components = X.shape[1]
+
+        if W is not None and H is not None:
+            n_samples_H = _num_samples(H)
+            n_features_W = _num_features(W)
+            if n_samples_H != n_features_W:
+                raise ValueError(
+                    "Incompatible shapes for H and W. Expected n_features for"
+                    f" W and n_samples for H to be equal. Got {n_samples_H} and"
+                    f" {n_features_W} instead."
+                )
```
So I think this check is not necessary anymore. the `_check_init` calls just below will complain if W and H don't have the appropriate shape.

Alexlandeau replied:
Deleted in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1352,9 +1397,14 @@ class NMF(_BaseNMF):
+    n_components : int or {'auto'} or None, default=None
         Number of components, if n_components is not set all features
         are kept.
+        If `n_components='auto'`, the number of components is automatically infered
+        from W or H shapes.
```
```suggestion
        If `n_components='auto'`, the number of components is automatically inferred
        from W or H shapes.
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1786,9 +1836,14 @@ class MiniBatchNMF(_BaseNMF):
+    n_components : int or {'auto'} or None, default=None
         Number of components, if `n_components` is not set all features
         are kept.
+        If `n_components='auto'`, the number of components is automatically infered
+        from W or H shapes.
```
```suggestion
        If `n_components='auto'`, the number of components is automatically inferred
        from W or H shapes.
```

Alexlandeau replied:
fixed in b3529bb1a0b276990b2124907d9b2f9b8f790b5d

jeremiedbb commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -932,3 +932,50 @@ def test_NMF_inverse_transform_W_deprecation():
+    assert H.shape[0] == 2
+    assert W.shape[1] == 2
+
+
+def test_nmf_n_components_default_value_warning():
```
```suggestion

# TODO(1.6): remove
def test_nmf_n_components_default_value_warning():
```

jeremiedbb commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -932,3 +932,50 @@ def test_NMF_inverse_transform_W_deprecation():
+    H_true = rng.random_sample((2, 5))
+    X = rng.random_sample((6, 5))
+    _, H, _ = non_negative_factorization(
+        X, H=H_true, n_components="auto", update_H=False
+    )  # should not fail
```
you can check that H.shape has not changed `H.shape == H_true.shape`.
you can also check that W has the appropriate shape.

jeremiedbb commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -932,3 +932,50 @@ def test_NMF_inverse_transform_W_deprecation():
+
+
+@pytest.mark.parametrize("Estimator", [NMF, MiniBatchNMF])
+def test_nmf_n_components_auto(Estimator):
+    rng = np.random.RandomState(0)
```
Please add a small description of the goal of this test, like "Check that n_components is correctly inferred from the provided custom initialization."

And similar for other tests

avigny replied:
done in https://github.com/scikit-learn/scikit-learn/pull/26634/commits/0e65acd8ff2ae46eb11dcfa5f5de5beb672302b2

jeremiedbb reviewed (commented):
Looking good. Here's a new batch of comments, mostly nitpicks.

jeremiedbb commented:
CI is failing because this PR introduces many FutureWarnings. You need to check every place where we call `(MiniBatch)NMF` or `non_negative_factorization` where `n_components` is left to its default value and change it to "auto" :)

Commit:
PR feedback: doc improvement and warning suppression

Commit:
merge conflicts

avigny committed:
Warn the user if `W` or `H` is set and will be overwritten at init

avigny committed:
do an actual test in test_nmf_n_components_auto_no_h_update

avigny committed:
update tests
adding descriptions at the begining of test functions
using less numerical values for assertion tests

avigny committed:
Update sklearn/decomposition/tests/test_nmf.py

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

avigny committed:
Better test comments

avigny committed:
Repairing docstring test

avigny committed:
Preventing n_components related warnings in tests

avigny committed:
update all usages of `non_negative_factorization`
with the `n_components="auto"` if needed to avoid the warnings

avigny commented:
Last thing, we also need to change the default value of `n_components` from `None` to `"warn"` in the `__init__` function of the two estimators.
this will likely create some (I mean a lot) of new warnings that we will have to  to take care of!

avigny committed:
set default value for n_components to `"warn"`

avigny committed:
Merge remote-tracking branch 'upstream/main' into bug/nmf_26392

avigny committed:
repair tests failing due to FutureWarning not caught

jeremiedbb commented:
Thinking about it again I think we should explicitly ignore the warnings instead of setting `n_components="auto"` in the tests. Sorry for my previous misleading comment. This can be done by decorating these tests with:
```
# TODO(1.6): remove the warning filter
@pytest.mark.filterwarnings("ignore:The default value of `n_components` will change")
```

avigny commented:
ok I'll do that!
TIL : adding `-Werror::FutureWarning` allows to have the same results as the CI but locally! great help from https://scikit-learn.org/dev/developers/tips.html#useful-pytest-aliases-and-flags

avigny committed:
Revert "repair tests failing due to FutureWarning not caught"

This reverts commit 918cdbe3047fe7b1f4880c801219da9d39971928.

avigny committed:
filter warnings related to n_components

avigny commented:
Should we also add `n_components="auto"` this example? I haven't found other places where the warning could occur
https://github.com/scikit-learn/scikit-learn/blob/bcbc2fc3abc32e879fba62486dabe9c05f615e79/examples/compose/plot_compare_reduction.py#L57

avigny committed:
test repair once again with Future Warning about n_components

avigny committed:
More precise changelog

avigny committed:
remove linebreak

jeremiedbb commented:
> Should we also add n_components="auto" this example? I haven't found other places where the warning could occur

Yes, we don't want warnings in the examples either

avigny commented:
> > Should we also add n_components="auto" this example? I haven't found other places where the warning could occur
> 
> Yes, we don't want warnings in the examples either

I was wrong, the `n_components` values are set by the `GridSearchCV` so there is no need to add a value for `n_components`
https://github.com/scikit-learn/scikit-learn/blob/bcbc2fc3abc32e879fba62486dabe9c05f615e79/examples/compose/plot_compare_reduction.py#L53-L70

jeremiedbb committed:
[doc build]

jeremiedbb reviewed (approved):
LGTM. Thanks @Alexlandeau and @avigny 

I fixed the remaining failing test and triggered the CI to run the examples to make sure we don't miss future warnings there. It'll be a longer run.

Alexlandeau committed:
Merge branch 'main' into bug/nmf_26392

Alexlandeau committed:
Merge branch 'main' into bug/nmf_26392

ogrisel commented on sklearn/decomposition/_nmf.py:
```diff
@@ -976,9 +979,14 @@ def non_negative_factorization(
+    n_components : int or {'auto'} or None, default=None
         Number of components, if n_components is not set all features
         are kept.
+        If `n_components='auto'`, the number of components is automatically inferred
+        from W or H shapes.
```
```suggestion
        If `n_components='auto'`, the number of components is
        automatically inferred from `W` or `H` shapes.
```

ogrisel commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -932,3 +954,88 @@ def test_NMF_inverse_transform_W_deprecation():
+
+
+def test_nmf_n_components_auto_no_h_update():
+    # Tests that non_negative_factorization does not fail
+    # when setting n_components="auto"
+    # also tests that the inferred n_component value is the right one
```
```suggestion
    # Tests that non_negative_factorization does not fail when setting
    # n_components="auto" also tests that the inferred n_component
    # value is the right one.
```

ogrisel reviewed (approved):
Once the missing test is added, LGTM.

avigny commented on sklearn/decomposition/_nmf.py:
```diff
@@ -1188,6 +1213,14 @@ def _check_params(self, X):
+            if H is not None:
+                self._n_components = _num_samples(H)
+            elif W is not None:
+                self._n_components = _num_features(W)
```
Small note about this branch:

if `H is None and W is not None` we now set `self._n_components = _num_features(W)` instead of `self._n_components = X.shape[1]`
#### Situation before to the PR
If you have `init = "custom"` ==> Error because `W` is not set
If you have `init != "custom"` ==> A new `W` is created with the shape `n_samples, X.shape[1]`. This is misleading because `X.shape[1]` can be different from `_num_features(W)`
see https://github.com/scikit-learn/scikit-learn/blob/9cbcc1f205e8be4dad1f383239e98381abb28bd0/sklearn/decomposition/_nmf.py#L1208-L1216

#### Situation after the PR
If you have `init = "custom"` [Same behaviour] ==> Error because `W` is not set
If you have `init != "custom"` [Changed behaviour]==> a new `W` is created with but now the shape of `W` will be the same as the shape of the `W` passed as argument

avigny replied:
the test was added in d8b6bf84ac93ef0b9fbd84cef45a0d12c0819705

jeremiedbb replied:
> If you have init != "custom" [Changed behaviour]==> a new W is created with but now the shape of W will be the same as the shape of the W passed as argument

I don't think this is a desirable behavior. It's also a breaking change. In that case we should completely ignore W and fallback to X.shape[1].

It requires a little bit of refactoring because it means that we can't compute n_components before checking in which situation we are. I directly pushed some changes to fix the behavior. The benefit is that we no longer have to rely on ``_num_samples`` and ``_num_features`` :)

Commit:
added test when W is used and not H

Alexlandeau committed:
Update sklearn/decomposition/tests/test_nmf.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Alexlandeau committed:
Update sklearn/decomposition/tests/test_nmf.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Alexlandeau committed:
Update sklearn/decomposition/tests/test_nmf.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Alexlandeau committed:
Update sklearn/decomposition/_nmf.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Alexlandeau committed:
Merge branch 'main' into bug/nmf_26392

Commit:
linting

Commit:
Merge branch 'bug/nmf_26392' of github.com:Alexlandeau/scikit-learn into bug/nmf_26392

jeremiedbb committed:
never use W to infer n_components when init!=custom

jeremiedbb committed:
improve coverage

Alexlandeau commented:
Thanks @jeremiedbb for the last review and suggestions !

avigny commented:
Thanks!!

yotamcons commented:
Thank you @Alexlandeau for this!
